packages/salesforcedx-vscode-core/src/settings/pushOrDeployOnSave.ts

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import { channelService } from '../channels';
+import { getWorkspaceOrgType, OrgType } from '../context';
+import { nls } from '../messages';
+import { notificationService } from '../notifications';
+import { sfdxCoreSettings } from '../settings';
+
+import { SfdxProjectJsonParser } from '@salesforce/salesforcedx-utils-vscode/out/src/cli';
+
+import * as fs from 'fs';
+import * as path from 'path';
+import * as vscode from 'vscode';
+
+export enum FileEventType {
+  Create,
+  Change,
+  Delete
+}
+
+const WAIT_TIME_IN_MS = 50;
+
+export async function registerPushOrDeployOnSave() {
+  if (sfdxCoreSettings.getPushOrDeployOnSaveEnabled()) {
```

Comment:
We only register the file watcher if the push-or-deploy-on-save setting is already set to true on core activation. This means that if a user wants to turn on the feature, they will have to restart VS Code. 

Disablement of the feature will be respected immediately will not require a restart of the extension.